src/MainWindow.vala

```diff
@@ -76,6 +76,16 @@ namespace PantheonCalculator {
             });
         }
 
+        public void undo () {
+            unowned List<History?> prev_entry = history.last ();
+            uint hist_length = history.length () - 1;
+            unowned List<History?> last_el = history.nth (hist_length);
+            if (prev_entry != null) {
+                entry.set_text (prev_entry.data.exp);
+            }
+            history.remove_link (last_el);
```

Comment:
If both of the elements are the same, you could remove `prev_entry` instead and move this line inside the if statement.